src/jit/flowgraph.cpp

```diff
@@ -21721,6 +21742,58 @@ void Compiler::fgDebugCheckBlockLinks()
     }
 }
 
+//------------------------------------------------------------------------------
+// fgDebugCheckNodesUniqueness: Check that each tree in the method has its own unique gtTreeId.
+//
```

Comment:
The cost of this method seems prohibitive, even for a checked build. It would seem that an approach that retains ranges of IDs that have been seen (e.g. using a B/B+ tree) would be preferable.

Reply:
This algorithm visit each node only once and it is linear in terms of trees.

Reply:
> This algorithm visits each node only once and it is linear in terms of trees.

Perhaps the bit-vector is not too prohibitive, given that it is linear cost. However, the algorithm, as far as I can tell, pushes *all* the nodes onto a stack and never pops them.
I don't think you need a stack at all for the LIR case - just do the check & add as you encounter each node. And in the non-LIR case you should pop & check & add the entire stack after each statement is handled.

Reply:
> However, the algorithm, as far as I can tell, pushes all the nodes onto a stack and never pops them

You are right, it is suffering from lack of std collections.
I can't pass bitVector to the visitor, without creating a decorator, because there is only one argument for the data.

Reply:
> because there is only one argument for the data.

I don't follow, what do you need more than one argument for?

Reply:
`BitVecOps::AddElemD(&nodesVecTraits, uniqueNodes, gtTreeID); `
we need the bit vector and its traits. 

I pushed a compromise, where we add trees for for one block each time, then check.

Reply:
Ah, traits, yes.

But I still don't understand why all this "complication" with the stack. Make a local struct/class. Put the bit vector and its traits in it. Pass its address to `fgWalkTreePre`. Done.

And you can put `fgMarkTreeId` in it to avoid adding more functions to `Compiler`. There are already way too many.

Reply:
> Make a local struct/class. Put the bit vector and its traits in it.

Yes, that's what most of the tree walkers do.

Reply:
Fixed, thanks. It looks much better now.